### ISSUE TITLE
redux, purity, concurrency

### ISSUE THREAD

**Issue by thedavidmeister** [holochain/holochain-rust#349]
i wanted to have a place to start distilling and recording the conversations we've had around this, probably splitting off child tickets and refactors

## scope of this ticket/discussion

- redux is a specific state management model developed in a single treaded environment (JS) by a company that is well known for working in single threaded environments (Facebook: PHP, JS)
  - redux does offer consistency of writes in the form of pure reducers against state values
  - redux does offer module decoupling in the form of arbitrary reducer dispatch against known actions
  - redux does offer co-ordination of writes in the form of the action queue/loop
  - redux does NOT NECESSARILY offer protection against things like [read skew](https://vladmihalcea.com/a-beginners-guide-to-read-and-write-skew-phenomena/) when several threads might make sequential reads (from their perspective) concurrently
  - redux COULD offer consistency guarantees across threads for interleaved reads and writes if we build in "as at" semantics into the state fetching (which would need structural sharing to do efficiently) or implement our functions that read from the redux store as pure, in the sense that the value of the store is passed into pure functions rather than allowing direct read access to the instance's state "now"
- we do NOT currently enforce pure reducers
  - this is BAD for concurrency and undermines the whole model
  - we currently have actions that `GetFoo`, which encourages non-purity and is BAD
- the naming of the redux store as "state" is confusing
  - from my personal background (clojure), i interpret "state" as "that which complects time and value"
    - using blockchain as an example, "value" is "get me the chain as at block X" and "state" is "get me the chain as at now"
  - as @lucksus pointed out, this is probably why redux calls it "store" and not "state" as the latter has a lot of conceptual baggage from functional world
  - a "store" _can_ be passed into pure functions and be handled robustly but reading directly from "state" immediately breaks purity
- given an action dispatch with multiple reducers, is the order of reducers logical and deterministic?
  - this is just a note to myself to find out :P

**Comment by thedavidmeister** [comment]
proposal for how to safely read from the store across threads:

don't do this:

```rust
fn foo(...) {
  let a = instance.state().some_module().some_result();
  // some write to the instance state can happen here causing a and b to see inconsistent values!!!
  // untestable and unreproduceable bugs...
  let b = instance.state().some_other_module().some_other_result();
}
```

instead do this:

```rust
fn foo(store: Store, ...) {
  let a = store.some_module().some_result();
  // even if the instance store processes a write here, the logic in this function is internally consistent
  // can be reliably unit tested, bugs are always reproduceable given a known Store value
  let b = store.some_other_module().some_other_result();
}

**Comment by thedavidmeister** [comment]
or alternatively, every new store value has an index/ID

```rust
fn foo(state_index: StateIndex, ...) {
  let a = instance.state(state_index).some_module().some_result();
  let b = instance.state(state_index).some_other_module().some_other_result();
}
```

i don't like this solution as much for a few reasons, but i thought i'd put it forward as a reference point for discussion

**Comment by lucksus** [comment]
The store is guarded by a RwLock, which means any access must go through either RwLock::read() or RwLock::write() which create RwLock***Guards. So if a reading context does not explicitly clone the read state content, throw away the lock and then use the clone, no write to the state can happen until the reader is complete and the lock goes out of scope.

In my action refactor PR I added Futures polling the state. In this line I'm getting that RwLockReadGuard: https://github.com/holochain/holochain-rust/pull/315/files#diff-d88a07eb05c270787f793360e568ce5aR47.

In order to have the state available there in the first place, I had to put a reference in the Context. But that reference is also wrapped in the RwLock: https://github.com/holochain/holochain-rust/blob/958907ee618c2bc3c6b9c01c4ea6989f0376b6de/core/src/context.rs#L47

**Comment by lucksus** [comment]
> redux does NOT NECESSARILY offer protection against things like read skew when several threads might make sequential reads (from their perspective) concurrently

That is a problem only if you treat an action solely as something setting a value. That would be a mistake. Actions should be thought of a high-level domain specific use-case. More like a transaction. Atomic state changes need to be expressed as an action and enforced by the reducer.

**Comment by lucksus** [comment]
I hope my sequence diagram on linking helps to shed some more light: https://hackmd.io/70sgYhlESsqv0QFrEgfZqw?view.

**Comment by lucksus** [comment]
As I state there in that diagram: we cannot have pure reducers as long as chain (and potentially DHT) are not part of the state but living in actors. That is why I said that IF we go with those actors, we need to couple them to the reducers (only change their state in a reducer, synchronously) so that they can still be regarded *logically* to be part of the state.

**Comment by ddd-mtl** [comment]
Whats kind of ironic in Nico's diagram is that the whole sequence is actually happening within a reduce of a ExecuteZomeFunction Action.
I don't see how a reduce can be pure if its waiting on nested action calls to be reduced.

**Comment by lucksus** [comment]
It is not. But currently it is a bit confusing. That's exactly what I'm working on right now with ticket #350. 

The code that runs in the thread of the ribosome and that is the calling context (through WASMI) of that sequence diagram is defined in a reducer right now, but since it's in a closure that gets executed in a separate thread, it is not happening in the context of that reducer. The reducer function has long returned the new state.. 

I am moving those asynchronous closures out of reducers and into action creators. And we should not put that in reducers going forward.

**Comment by ddd-mtl** [comment]
hmmm okay. I think it would help to understand if your diagram showed the ExecuteZomeFunction action that initiated the invoke_commit_link(). Or maybe another diagram with just the ExecuteZomeFunction sequence?

**Comment by Tails** [comment]
So, does the fact that this ticket is open mean writing entries can not yet be done atomically? Context for question: https://forum.holochain.org/t/threading-thread-safety/3963/3

**Comment by thedavidmeister** [comment]
@Tails all wasm calls are atomic now
